exchanges/kucoin/kucoin_websocket.go

```diff
@@ -977,396 +981,197 @@ func (ku *Kucoin) processMarketSnapshot(respData []byte, instrument string) erro
-			futuresContractMarketDataChannel,
-			futuresTradeOrdersBySymbolChannel,
-			futuresPositionChangeEventChannel,
-			futuresTrasactionStatisticsTimerEventChannel:
-			for b := range futuresPairs {
-				futuresPairs[b], err = ku.FormatExchangeCurrency(futuresPairs[b], asset.Futures)
+			subs := spotOrMarginPairSubs(assetPairs, s, false, interval)
+			subscriptions = append(subscriptions, subs...)
+		case s.Channel == marginFundingbookChangeChannel:
+			s.Channel = fmt.Sprintf(s.Channel, assetPairs[asset.Margin].GetCurrencies().Join())
+			subscriptions = append(subscriptions, *s)
+		case s.Channel == marketSnapshotChannel:
+			subs := spotOrMarginCurrencySubs(assetPairs, s)
+			subscriptions = append(subscriptions, subs...)
+		case getChannelsAssetType(s.Channel) == asset.Futures && isSymbolChannel(s.Channel):
+			for _, p := range assetPairs[asset.Futures] {
+				c, err := ku.FormatExchangeCurrency(p, asset.Futures)
 				if err != nil {
 					continue
 				}
-				subscriptions = append(subscriptions, stream.ChannelSubscription{
-					Channel:  channels[x],
-					Asset:    asset.Futures,
-					Currency: futuresPairs[b],
-				})
+				s := *s
+				s.Channel = fmt.Sprintf(s.Channel, c)
+				subscriptions = append(subscriptions, s)
 			}
+		case isSymbolChannel(s.Channel):
+			// Subscriptions which can use a single comma-separated sub per asset
+			subs := spotOrMarginPairSubs(assetPairs, s, true)
+			subscriptions = append(subscriptions, subs...)
+		default:
+			subscriptions = append(subscriptions, *s)
 		}
 	}
 	return subscriptions, nil
 }
 
-func (ku *Kucoin) generatePayloads(subscriptions []stream.ChannelSubscription, operation string) ([]WsSubscriptionInput, error) {
-	payloads := make([]WsSubscriptionInput, 0, len(subscriptions))
-	marketTickerSnapshotForCurrencyChannelCurrencyFilter := map[currency.Code]int{}
-	for x := range subscriptions {
-		var err error
-		var a asset.Item
-		a, err = ku.getChannelsAssetType(subscriptions[x].Channel)
-		if err != nil {
-			return nil, err
+// isSymbolChannel returns true it this channel path ends in a formatting %s to accept a Symbol
+func isSymbolChannel(c string) bool {
+	return strings.HasSuffix(c, "%s") || strings.HasSuffix(c, "%v")
+}
+
+// channelName converts global channel Names used in config of channel input into kucoin channel names
+// returns the name unchanged if no match is found
+func channelName(name string) string {
+	if s, ok := subscriptionNames[name]; ok {
+		return s
+	}
+	return name
+}
+
+// spotOrMarginPairSubs accepts a map of pairs and a template subscription and returns a list of subscriptions for Spot and Margin pairs
+// If there's a Spot subscription, it won't be added again as a Margin subscription
+// If joined param is true then one subscription per asset type with the currencies comma delimited
+func spotOrMarginPairSubs(assetPairs map[asset.Item]currency.Pairs, base *subscription.Subscription, join bool, fmtArgs ...any) []subscription.Subscription {
```

Comment:
base name conflicts with a bunch of base packages. If you could rename that, that would be good,

Reply:
Yep, thanks 🙏 
Fixed a1f316a